cloud,bulk: include external IO requests in tenant usage

stevendanna opened the pull request:
This adds basic accounting to ExternalStorage writes. The measured
bytes written is almost surely inaccurate as it doesn't account for

- retries inside client libraries,
- protocol metadata such as HTTP headers, or
- transport level compression.

Thus, instead of measuring "bytes sent over the network" it is
measuring something closer to "bytes stored on external storage
services"

Release justification: Low risk (off by default) feature for multi-tenant
cost accounting.

Release note: Tenant usage now includes bytes written to external data
storage services such as Google Compute Storage or Amazon S3.

stevendanna committed:
bulk,cdc: account for external IO requests in tenant usage

This adds new methods to the tenant-side cost collector that records
RU usage for "external io" requests.

This accounting is present by default in all external storage
instances constructed using the constructors that live on
ExecCfg. Callers can opt out of the accounting by passing their own
ReadWriteInterceptor. For example, changefeeds do this as they record
their usage centrally in pre-existing infrastructure to record
metrics.

Currently, the newly added TenantSideExternalIORecorder will not block
or increment RUs. However, this behaviour is configurable via the
tenant_external_io_ru_accounting_mode cluster setting.

Release note (ops change): Tenants now account for bytes read and
written to external services. Such reads and writes consume tenant
resource units.

Release justification: High priority feature for multi-tenant resource
accounting.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/76495)
<!-- Reviewable:end -->

dt commented:
I haven't given this a detailed read yet, but my first reaction is to wonder if we need to push this into the pkg/cloud API this deeply, or if the caller could a) use the existing cloud API and then b) turn around and talk to a cost controller, potentially with a util/helper. I guess pushing it into the cloud API itself makes it easier to account mid-operation, in every Write call, like this, while a simplistic version of caller-driven accounting a layer up would probably start with just accounting after the file is finished? Though even that doesn't seem forced -- we could easily have a wrapper in the cost control library that the caller applies to the io.Writer it gets from pkg/cloud, if we really way per-buf accounting vs per-file?

Also, we want to account for reads too, right?

stevendanna commented:
> I haven't given this a detailed read yet, but my first reaction is to wonder if we need to push this into the pkg/cloud API this deeply, or if the caller could a) use the existing cloud API and then b) turn around and talk to a cost controller, potentially with a util/helper.

Yeah, this is what I went back and forth on this morning.  I'll try to give it some more thought today.  A thing I like about adding the accounting to the io.Writer's returned by the ExternalStorage is that the the callers largely don't need to care; but, as with anything, with the correct helper, I'm sure we could get a similar effect by caller-driven accounting as well.

> Also, we want to account for reads too, right?

I believe we will want to account for both. I am unsure if we want to account for some way for other operations. For example, I believe `List()` operations cost money on s3.

RaduBerinde reviewed (commented):
Good job figuring out how to change the RU subsystems, the approach looks good.

We are trying to account for data transfer costs, correct? Then the model should be like the "pgwire egress", where we just have a bytes cost (and not like the KV operations model where we also have a per-request cost). Feels like we'd need something like "external storage egress" and "external storage ingress", no?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76495#-:-MvtBZ1r5-1-99roBOyS:b-vz7cte)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @stevendanna)

___
*[pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go, line 625 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76495#-MvtCQcZ4wEORRlJa1pg:-MvtCQcZ4wEORRlJa1ph:b-c3ub8n) ([raw file](https://github.com/cockroachdb/cockroach/blob/62fdcc48e0219fd2c29469a5c950a44844ce6131/pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go#L625)):*
> ```Go
> 
> // OnExternalWriteResponse is part of the multitenant.TenantSideExternalIOInterceptor interface.
> func (c *tenantSideCostController) OnExternalWriteResponse(ctx context.Context, bytes int64) {
> ```

[nit] I'd name it something like `RecordExternalStorageIngress`, since it's not an on-request/on-response pair.


<!-- Sent from Reviewable.io -->

stevendanna commented:
@dt @miretskiy I've still got more tests to add but I was wondering if you could take look at the approach here.  I've tried to balance a few of the tensions: We don't want to involve ourselves in RU accounting without also blocking for RUs, but we also don't want to block in places where it would cause new performance problems and we also want to be able to make local decisions about how we account for reads/writes.

Here, I've moved the accounting out of the cloud API, but still provide a wrapper ExternalStorage for cases where that seemed to me like what we wanted to do.  The wrapping external storage now returns readers/writers that keep a local count that they only send to the tenantcostserver periodically (every N bytes) and on Close().  

The API for recording write costs also knows how to account for failed writes that didn't write as much as they thought they were going to.

Wiring this up is a bit more "shotgun" than I would like currently, but I wanted to be conservative about where we added this accounting for now and just hit the biggest offenders.

I'm not particularly attached to any of this, but I am attached to at least getting _some_ level of accounting in that we feel is safe.

miretskiy commented on pkg/cloud/gcp/gcs_storage_test.go:
```diff
@@ -173,9 +173,9 @@ func TestCompressedGCS(t *testing.T) {
 	require.NoError(t, err)
 
-	s1, err := cloud.MakeExternalStorage(ctx, conf1, base.ExternalIODirConfig{}, testSettings, nil, nil, nil)
+	s1, err := cloud.MakeExternalStorage(ctx, conf1, base.ExternalIODirConfig{}, testSettings, nil, nil, nil, nil)
```
would be nice to `require.True(t, s1.RequiresExternalIOAccounting())` (in other implementation tests too)

miretskiy commented on pkg/multitenant/cost_controller.go:
```diff
@@ -96,6 +98,46 @@ func HasTenantCostControlExemption(ctx context.Context) bool {
+	ExternalIOWriteFailure(ctx context.Context, used int64, returned int64)
+
+	// RecordExternalStroageEgressSuccess records the usage of RUs associated with
+	// a write of the given size. This function does not wait.
```
stray comment or missing method?

stevendanna replied:
These comments were all for typo'd versions of the old name. Should be fixed now.

miretskiy commented on pkg/multitenant/cost_controlling_io.go:
```diff
@@ -0,0 +1,208 @@
+// DefaultBytesAllowedBeforeAccounting are how maybe bytes we will read.write before trying to wait for
+// RUs. The goal here is to avoid waiting in loops for the common case without allowing an
+// unbounded number of bytes read/written before accounting for them.
+const DefaultBytesAllowedBeforeAccounting = 128 << 20 // 128 MB
```
does this deserve to be a setting ?

dt replied:
or an envutil.WithDefault

128mb seems high to me -- I'd have expected 16 or less?

miretskiy commented on pkg/multitenant/cost_controlling_io.go:
```diff
@@ -0,0 +1,208 @@
+	limit    int64
+}
+
+func NewExternalStorageWithAccounting(
```
exported functions need comment?

miretskiy commented on pkg/multitenant/cost_controlling_io.go:
```diff
@@ -0,0 +1,208 @@
+// Close closes the underlying Writer and also waits for any RUs that weren't accounted for on a
+// previous call to Write.
+func (aw *accountingWriter) Close() error {
+	aw.recorder.ExternalIOWriteSuccess(aw.ctx, aw.count)
```
a bit strange seeing "WriteSuccess" here because close should be invoked even if write failed.

stevendanna replied:
I'll add a comment to point out that `count` are the bytes that we've previously successfully written.

miretskiy commented on pkg/ccl/changefeedccl/metrics.go:
```diff
@@ -62,6 +64,20 @@ type AggMetrics struct {
 
+type metricsRecorderBuilder func(requiresCostAccounting bool) metricsRecorder
+
+type metricsRecorder interface {
```
nice. A demonstration of an adage that "all problem can be solved by indirection".

miretskiy commented on pkg/ccl/changefeedccl/metrics.go:
```diff
@@ -174,6 +190,57 @@ func (m *sliMetrics) getBackfillRangeCallback() func(int64) (func(), func()) {
+	// NB: We don't Wait for RUs for changefeeds; but, this call may put the RU limiter in debt which
+	// will impact future KV requests.
+	w.recorder.ExternalIOWriteSuccess(w.ctx, int64(compressedBytes))
+}
```
nit: add "\n"?

miretskiy commented on pkg/ccl/changefeedccl/metrics.go:
```diff
@@ -174,6 +190,57 @@ func (m *sliMetrics) getBackfillRangeCallback() func(int64) (func(), func()) {
+	innerCallback := w.inner.recordEmittedMessages()
+	return func(numMessages int, mvcc hlc.Timestamp, bytes int, compressedBytes int) {
+		w.recordEmittedBatch(time.Time{}, numMessages, mvcc, bytes, compressedBytes)
+		innerCallback(numMessages, mvcc, bytes, compressedBytes)
```
this is actually very elegant -- works even for kafka

miretskiy commented on pkg/ccl/changefeedccl/metrics.go:
```diff
@@ -174,6 +190,57 @@ func (m *sliMetrics) getBackfillRangeCallback() func(int64) (func(), func()) {
 
+type wrappingCostController struct {
+	ctx      context.Context
+	inner    metricsRecorder
```
if you embed inner recorder, you won't have to override noop methods below -- just the messages callback.

stevendanna replied:
I didn't do this on purpose so that we have to decide explicitly when new methods are added whether or not they need accounting.

miretskiy commented on pkg/ccl/changefeedccl/sink_kafka.go:
```diff
@@ -669,7 +669,7 @@ func makeKafkaSink(
 		bootstrapAddrs: u.Host,
 		topics:         makeTopicsMap(kafkaTopicPrefix, kafkaTopicName, targets),
-		metrics:        m,
+		metrics:        mb(true),
```
nit: add /* */ comment for true.  Better yet:
```
const requireResourceAccounting = true
```

miretskiy commented on pkg/ccl/changefeedccl/sink_sql.go:
```diff
@@ -106,7 +109,7 @@ func makeSQLSink(
 		hasher:      fnv.New32a(),
 		targetNames: targetNames,
-		metrics:     m,
+		metrics:     mb(false),
```
wouldn't that depend where the sql traffic going?

stevendanna replied:
Yes, but adding a reliable way of knowing where the sql traffic was going seemed like it deserves its own thought so for now I think we should error in favor of the user.  I'll add a comment.

miretskiy commented on pkg/ccl/changefeedccl/sink_webhook.go:
```diff
@@ -300,7 +300,7 @@ func makeWebhookSink(
 		parallelism: parallelism,
 		ts:          source,
-		metrics:     m,
+		metrics:     mb(true),
```
could be localhost, but I guess that's fine.
I commented elsewhere, but I think it would be nice to add constants to require/opt-out of resource acccounting.

miretskiy commented on pkg/ccl/changefeedccl/testing_knobs.go:
```diff
@@ -34,6 +34,8 @@ type TestingKnobs struct {
 	// FeedKnobs are kvfeed testing knobs.
 	FeedKnobs kvfeed.TestingKnobs
+
+	NullSinkIsExternalIOAccounted bool
```
nit: comment?

miretskiy commented on pkg/ccl/backupccl/backup_job.go:
```diff
@@ -627,7 +627,6 @@ func (b *backupResumer) Resume(ctx context.Context, execCtx interface{}) error {
 			return err
 		}
 		defer c.Close()
-
```
probably shouldn't be included in this pr?

miretskiy replied:
(but... you know... it's fine if it is.

miretskiy commented on pkg/multitenant/cost_controlling_io.go:
```diff
@@ -0,0 +1,208 @@
+
+func (aw *accountingWriter) immediatelyAccountedWrite(d []byte) (int, error) {
+	writeLen := int64(len(d))
+	aw.recorder.ExternalIOWriteWait(aw.ctx, writeLen)
```
if writeLen > that the overall limit in that quota pool?

stevendanna replied:
I believe this limiter should have a steady flow of new capacity added to it, so it will eventually return IIRC.

miretskiy reviewed (commented):
Took first pass -- looks reasonable, but I'll need to look more carefully at accounting readers/writes.
Changefeed specific changes look very nice.

stevendanna commented:
@dt Wondering your thoughts on keeping this opt-in at a few call sites (as it is now) vs going back to having it enabled by default for ExternalStorage's constructed in tenants (with then an option to opt-out at call sites that want to do the accounting outside of the readers and writers).

I've set RU accounting to `off` for now to make this a bit safer to merge.

dt commented:
Yeah, I'm a little uneasy about all the opt-in call-sites and missing things/becoming inconsistent.

We have the accounting wrapper. Almost all users get an implementation instance by calling the factory field of evalctx that we setup in pkg/server. Could we alter that factory to return wrapped instances for MT callers, so that they get automatic accounting, when they use the object they get back from `flowCtx.EvalCtx.ExternalStorage(...)` as-is, with no changes at the call-sites, vs having to call things like `WriteFileAccounted()` ?

stevendanna commented:
> We have the accounting wrapper. Almost all users get an implementation instance by calling the factory field of evalctx that we setup in pkg/server. Could we alter that factory to return wrapped instances for MT callers, so that they get automatic accounting, when they use the object they get back from flowCtx.EvalCtx.ExternalStorage(...) as-is, with no changes at the call-sites, vs having to call things like WriteFileAccounted() ?

@dt Happy to, should be pushed now.

dt reviewed (commented):
I like the pkg/cloud API changes a lot better this way -- with the factory applying the options -- I think? I mostly only looked at those pkg/cloud API changes and their usage/forced changes in the bulk cod while briefly skimming the tenantcost* stuff, but overall LGTM.

I'd feel _slightly_ better if we wrote a microbenchmark of http sink (to a local in-process server) and tested read and write throughput with and without both accounting and ru limiting enabled, but I won't block merging this for that.

stevendanna commented:
The following are some microbenchmarks of 8 goroutines writing 8MB files to a null sink with various ru-accounting and limit settings and a egress cost of 1 RU per MB.  As expected, there is some overhead even when ru-accounting is set to nowait.

```
BenchmarkExternalIOAccounting/op=write/without-interceptor
BenchmarkExternalIOAccounting/op=write/without-interceptor/concurrency=8
BenchmarkExternalIOAccounting/op=write/without-interceptor/concurrency=8-16               230460              4781 ns/op            1157 B/op         27 allocs/op
BenchmarkExternalIOAccounting/op=write/with-interceptor
BenchmarkExternalIOAccounting/op=write/with-interceptor/ru-accounting=on/limit=1024/concurrency=8
BenchmarkExternalIOAccounting/op=write/with-interceptor/ru-accounting=on/limit=1024/concurrency=8-16               58702             54036 ns/op            2961 B/op         47 allocs/op
BenchmarkExternalIOAccounting/op=write/with-interceptor/ru-accounting=on/limit=16384/concurrency=8
BenchmarkExternalIOAccounting/op=write/with-interceptor/ru-accounting=on/limit=16384/concurrency=8-16              14492             84121 ns/op            2296 B/op         44 allocs/op
BenchmarkExternalIOAccounting/op=write/with-interceptor/ru-accounting=on/limit=16777216/concurrency=8
BenchmarkExternalIOAccounting/op=write/with-interceptor/ru-accounting=on/limit=16777216/concurrency=8-16           14467             84061 ns/op            2365 B/op         45 allocs/op
BenchmarkExternalIOAccounting/op=write/with-interceptor/ru-accounting=off/limit=1024/concurrency=8
BenchmarkExternalIOAccounting/op=write/with-interceptor/ru-accounting=off/limit=1024/concurrency=8-16             169297              7192 ns/op            1753 B/op         35 allocs/op
BenchmarkExternalIOAccounting/op=write/with-interceptor/ru-accounting=off/limit=16384/concurrency=8
BenchmarkExternalIOAccounting/op=write/with-interceptor/ru-accounting=off/limit=16384/concurrency=8-16            170145              6839 ns/op            1658 B/op         35 allocs/op
BenchmarkExternalIOAccounting/op=write/with-interceptor/ru-accounting=off/limit=16777216/concurrency=8
BenchmarkExternalIOAccounting/op=write/with-interceptor/ru-accounting=off/limit=16777216/concurrency=8-16         195210              5928 ns/op            1654 B/op         35 allocs/op
BenchmarkExternalIOAccounting/op=write/with-interceptor/ru-accounting=nowait/limit=1024/concurrency=8
BenchmarkExternalIOAccounting/op=write/with-interceptor/ru-accounting=nowait/limit=1024/concurrency=8-16          169041              6885 ns/op            1680 B/op         35 allocs/op
BenchmarkExternalIOAccounting/op=write/with-interceptor/ru-accounting=nowait/limit=16384/concurrency=8
BenchmarkExternalIOAccounting/op=write/with-interceptor/ru-accounting=nowait/limit=16384/concurrency=8-16         162897              6936 ns/op            1652 B/op         35 allocs/op
BenchmarkExternalIOAccounting/op=write/with-interceptor/ru-accounting=nowait/limit=16777216/concurrency=8
BenchmarkExternalIOAccounting/op=write/with-interceptor/ru-accounting=nowait/limit=16777216/concurrency=8-16      191762              6066 ns/op            1646 B/op         35 allocs/op
```

The following read uses an in-memory http server:

```
BenchmarkExternalIOAccounting/op=read/without-interceptor
BenchmarkExternalIOAccounting/op=read/without-interceptor/concurrency=8
BenchmarkExternalIOAccounting/op=read/without-interceptor/concurrency=8-16                    72          14321238 ns/op          219144 B/op       1464 allocs/op
BenchmarkExternalIOAccounting/op=read/with-interceptor
BenchmarkExternalIOAccounting/op=read/with-interceptor/ru-accounting=on/limit=1024/concurrency=8
BenchmarkExternalIOAccounting/op=read/with-interceptor/ru-accounting=on/limit=1024/concurrency=8-16                   84          13886783 ns/op          172049 B/op       1017 allocs/op
BenchmarkExternalIOAccounting/op=read/with-interceptor/ru-accounting=on/limit=16384/concurrency=8
BenchmarkExternalIOAccounting/op=read/with-interceptor/ru-accounting=on/limit=16384/concurrency=8-16                  84          14294107 ns/op          531463 B/op       3849 allocs/op
BenchmarkExternalIOAccounting/op=read/with-interceptor/ru-accounting=on/limit=16777216/concurrency=8
BenchmarkExternalIOAccounting/op=read/with-interceptor/ru-accounting=on/limit=16777216/concurrency=8-16               73          14422466 ns/op          196070 B/op       1117 allocs/op
BenchmarkExternalIOAccounting/op=read/with-interceptor/ru-accounting=off/limit=1024/concurrency=8
BenchmarkExternalIOAccounting/op=read/with-interceptor/ru-accounting=off/limit=1024/concurrency=8-16                  81          14054741 ns/op          182530 B/op       1065 allocs/op
BenchmarkExternalIOAccounting/op=read/with-interceptor/ru-accounting=off/limit=16384/concurrency=8
BenchmarkExternalIOAccounting/op=read/with-interceptor/ru-accounting=off/limit=16384/concurrency=8-16                 81          13901101 ns/op          172371 B/op       1019 allocs/op
BenchmarkExternalIOAccounting/op=read/with-interceptor/ru-accounting=off/limit=16777216/concurrency=8
BenchmarkExternalIOAccounting/op=read/with-interceptor/ru-accounting=off/limit=16777216/concurrency=8-16              76          14041702 ns/op          175238 B/op       1007 allocs/op
BenchmarkExternalIOAccounting/op=read/with-interceptor/ru-accounting=nowait/limit=1024/concurrency=8
BenchmarkExternalIOAccounting/op=read/with-interceptor/ru-accounting=nowait/limit=1024/concurrency=8-16               80          14449179 ns/op          153829 B/op       1057 allocs/op
BenchmarkExternalIOAccounting/op=read/with-interceptor/ru-accounting=nowait/limit=16384/concurrency=8
BenchmarkExternalIOAccounting/op=read/with-interceptor/ru-accounting=nowait/limit=16384/concurrency=8-16              84          14946318 ns/op          333819 B/op       1372 allocs/op
BenchmarkExternalIOAccounting/op=read/with-interceptor/ru-accounting=nowait/limit=16777216/concurrency=8
BenchmarkExternalIOAccounting/op=read/with-interceptor/ru-accounting=nowait/limit=16777216/concurrency=8-16           70          15850243 ns/op          194913 B/op       1068 allocs/op
```

dt commented:
> The following read uses an in-memory http server

Do we have a sense for why the hit on reads is *so* much worse?

stevendanna commented:
> Do we have a sense for why the hit on reads is so much worse?

I think this test ends up mostly benchmarking the in memory http server. For instance, in the data pasted above, some of the benchmarks for the interceptor are faster than without the interceptor.

stevendanna commented:
@dt This is just rebased on top of your limiter worker. The main change is that we now always wrap the external storage in a wrapper that adds either or both of our reader/writer implementations.

stevendanna commented:
bors r=dt
